internal/topology/names/names.go

```diff
@@ -0,0 +1,127 @@
+)
+
+const (
+	maxNameLength          = 63
+	randomLength           = 5
+	maxGeneratedNameLength = maxNameLength - randomLength
+)
+
+type simpleNameGenerator struct {
+	base string
+}
+
+func (s *simpleNameGenerator) GenerateName() (string, error) {
+	base := s.base
+	if len(base) > maxGeneratedNameLength {
+		base = base[:maxGeneratedNameLength]
+	}
+	return fmt.Sprintf("%s%s", base, utilrand.String(randomLength)), nil
+}
+
+// NameGenerator generates names for objects.
+type NameGenerator interface {
+	// GenerateName generates a valid name. The generator is responsible for
+	// knowing the maximum valid name length.
+	GenerateName() (string, error)
+}
+
+// SimpleNameGenerator returns a NameGenerator which is based on
+// k8s.io/apiserver/pkg/storage/names.SimpleNameGenerator.
+func SimpleNameGenerator(base string) NameGenerator {
+	return &simpleNameGenerator{
+		base: base,
+	}
+}
+
+// ControlPlaneNameGenerator returns a generator for creating a control plane name.
+func ControlPlaneNameGenerator(templateString, clusterName string) NameGenerator {
+	return newTemplateGenerator(templateString, clusterName,
+		map[string]interface{}{})
+}
+
+// MachineDeploymentNameGenerator returns a generator for creating a machinedeployment name.
+func MachineDeploymentNameGenerator(templateString, clusterName, topologyName string) NameGenerator {
+	return newTemplateGenerator(templateString, clusterName,
+		map[string]interface{}{
+			"machineDeployment": map[string]interface{}{
+				"topologyName": topologyName,
+			},
+		})
+}
+
+// MachinePoolNameGenerator returns a generator for creating a machinepool name.
+func MachinePoolNameGenerator(templateString, clusterName, topologyName string) NameGenerator {
+	return newTemplateGenerator(templateString, clusterName,
+		map[string]interface{}{
+			"machinePool": map[string]interface{}{
+				"topologyName": topologyName,
+			},
+		})
+}
+
+// templateGenerator parses the template string as text/template and executes it using
+// the passed data to generate a name.
+type templateGenerator struct {
+	template string
+	data     map[string]interface{}
+}
+
+func newTemplateGenerator(template, clusterName string, data map[string]interface{}) NameGenerator {
+	data["cluster"] = map[string]interface{}{
+		"name": clusterName,
+	}
+	data["random"] = utilrand.String(randomLength)
+
+	return &templateGenerator{
+		template: template,
+		data:     data,
+	}
+}
+
+func (g *templateGenerator) GenerateName() (string, error) {
+	tpl, err := template.New("template name generator").Option("missingkey=error").Parse(g.template)
+	if err != nil {
+		return "", errors.Wrapf(err, "parsing template %q", g.template)
+	}
+
+	var buf bytes.Buffer
+	if err := tpl.Execute(&buf, g.data); err != nil {
+		return "", errors.Wrapf(err, "rendering template: %q", tpl.Name())
+	}
+
+	name := buf.String()
+
+	// If the name exceeds the maxNameLength: trim to maxGeneratedNameLength and add
+	// a random suffix.
+	if len(name) > maxNameLength {
```

Comment:
```suggestion
	if len(name) > maxGeneratedNameLength {
```

I wonder if we should use this for consistency. Just seems strange to use a different constant here compared to above

Reply:
Otherwise we should probably use the same in SimpleNameGenerator but that would diverge from the upstream SimpleNameGenerator

Reply:
If we change this, we should also update the godoc on the API

Reply:
It is like this on purpose!

SimpleNameGenerator **always** adds a random suffix of length 5. So the input base string is not allowed to be longer than `maxGeneratedNameLength` (58), which results in a name of maximum `maxNameLength` (63).

The templateGenerator **only** trims and adds a random suffix of length 5, if the generated name exceeds the `maxNameLength` (63).

Reply:
Note: the `{{ .random }}` could be anywhere in the template.

Reply:
Yup. I assumed it was on purpose, but it's still an inconsistency between the generators :)